src/main/java/org/mockito/internal/creation/bytebuddy/SubclassBytecodeGenerator.java

```diff
@@ -97,10 +98,31 @@ private static boolean needsSamePackageClassLoader(MockFeatures<?> features) {
             // package private methods.
             return true;
         }
+        if (hasPackagePrivateParam(features.mockedType)) {
+            return true;
+        }
+
         for (Class<?> iface : features.interfaces) {
             if (!Modifier.isPublic(iface.getModifiers())) {
                 return true;
             }
+            if (hasPackagePrivateParam(iface)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static boolean hasPackagePrivateParam(Class<?> iface) {
```

Comment:
Nit: `hasPackagePrivateTypeReference`, since we are also checking the return type.

Reply:
Renamed to hasNonPublicTypeReference